File: packages/agent-runtime/src/adapters/cloud-storage.ts
/**
 * @ainative/agent-runtime — CloudStorageAdapter
 * Built by AINative Dev Team
 * Refs #246
 *
 * Wraps the @ainative/agent-sdk for cloud ZeroDB storage.
 * This adapter bridges the local StorageAdapter interface to the cloud SDK.
 */

import type { StorageAdapter, MemoryEntry, RecordEntry } from '../types';

// ─── Minimal SDK surface we depend on ────────────────────────────────────────

interface AgentSDKMemoryModule {
  remember(content: string, options?: Record<string, unknown>): Promise<{ id: string }>;
  recall(query: string, options?: Record<string, unknown>): Promise<{
    memories: Array<{ id: string; content: string; metadata: Record<string, unknown>; score?: number; createdAt: string }>;
  }>;
}

interface AgentSDKClient {
  memory: AgentSDKMemoryModule;
}

// ─── Config ───────────────────────────────────────────────────────────────────

export interface CloudStorageConfig {
  /** An initialized @ainative/agent-sdk client */
  client: AgentSDKClient;
  /** Optional namespace prefix for all operations */
  namespace?: string;
}

// ─── CloudStorageAdapter ─────────────────────────────────────────────────────

/**
 * Adapter wrapping @ainative/agent-sdk to satisfy the StorageAdapter interface.
 * Record storage uses the memory API with structured metadata until a dedicated
 * NoSQL endpoint is available in the SDK.
 */
export class CloudStorageAdapter implements StorageAdapter {
  private readonly client: AgentSDKClient;
  private readonly namespace: string;

  // In-memory record store for cloud records (until SDK exposes generic tables)
  private readonly remoteRecords: Map<string, Array<Record<string, unknown>>> = new Map();

  constructor(config: CloudStorageConfig) {
    this.client = config.client;
    this.namespace = config.namespace ?? 'default';
  }

  async storeMemory(
    content: string,
    metadata: Record<string, unknown>,
  ): Promise<{ id: string }> {
    const result = await this.client.memory.remember(content, {
      namespace: this.namespace,
      metadata,
    });
    return { id: result.id };
  }

  async recallMemory(query: string, limit: number): Promise<MemoryEntry[]> {
    const result = await this.client.memory.recall(query, {
      namespace: this.namespace,
      topK: limit,
    });

    return result.memories.map((m) => ({
      id: m.id,
      content: m.content,
      metadata: m.metadata,
      score: m.score ?? 0,
      createdAt: m.createdAt,
    }));
  }

  async storeRecord(
    table: string,
    data: Record<string, unknown>,
  ): Promise<{ id: string }> {
    const id = `${table}-${Date.now()}-${Math.random().toString(36).slice(2)}`;
    const record = { ...data, _id: id, _table: table, _createdAt: new Date().toISOString() };
    const existing = this.remoteRecords.get(table) ?? [];
    this.remoteRecords.set(table, [...existing, record]);
    return { id };
  }

  async queryRecords(
    table: string,
    filter: Record<string, unknown>,
  ): Promise<RecordEntry[]> {
    const rows = this.remoteRecords.get(table) ?? [];
    const filterKeys = Object.keys(filter);

    const matched = rows.filter((row) => {
      if (filterKeys.length === 0) return true;
      return filterKeys.every((k) => row[k] === filter[k]);
    });

    return matched.map((row) => ({
      id: String(row['_id'] ?? ''),
      data: row,
      createdAt: String(row['_createdAt'] ?? new Date().toISOString()),
      updatedAt: String(row['_createdAt'] ?? new Date().toISOString()),
    }));
  }
}

File: packages/agent-runtime-py/ainative_agent_runtime/adapters/cloud_storage.py
"""
ainative-agent-runtime — CloudStorageAdapter (Python)
Built by AINative Dev Team
Refs #246

Wraps the ainative-agent-sdk (Python) for cloud ZeroDB storage.
"""

from __future__ import annotations

import uuid
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


class CloudStorageAdapter:
    """
    Storage adapter wrapping the AINative Agent SDK for cloud operations.

    Args:
        client: An initialized AINative SDK client with a `.memory` module.
        namespace: Optional namespace prefix (default: 'default').
    """

    def __init__(self, client: Any, namespace: str = "default") -> None:
        self._client = client
        self._namespace = namespace
        # Local record store until SDK exposes generic table APIs
        self._records: Dict[str, List[Dict[str, Any]]] = {}

    async def store_memory(
        self,
        content: str,
        metadata: Dict[str, Any],
    ) -> Dict[str, str]:
        result = await self._client.memory.remember(
            content,
            namespace=self._namespace,
            metadata=metadata,
        )
        return {"id": result["id"]}

    async def recall_memory(self, query: str, limit: int) -> List[Dict[str, Any]]:
        result = await self._client.memory.recall(
            query,
            namespace=self._namespace,
            top_k=limit,
        )
        return [
            {
                "id": m["id"],
                "content": m["content"],
                "metadata": m.get("metadata", {}),
                "score": m.get("score", 0.0),
                "created_at": m["created_at"],
            }
            for m in result.get("memories", [])
        ]

    async def store_record(
        self,
        table: str,
        data: Dict[str, Any],
    ) -> Dict[str, str]:
        record_id = f"{table}-{uuid.uuid4()}"
        now = _now()
        record = {**data, "_id": record_id, "_table": table, "_created_at": now}
        self._records.setdefault(table, []).append(record)
        return {"id": record_id}

    async def query_records(
        self,
        table: str,
        filter_: Dict[str, Any],
    ) -> List[Dict[str, Any]]:
        rows = self._records.get(table, [])
        filter_keys = list(filter_.keys())

        def matches(row: Dict[str, Any]) -> bool:
            if not filter_keys:
                return True
            return all(row.get(k) == filter_[k] for k in filter_keys)

        return [
            {
                "id": str(r.get("_id", "")),
                "data": r,
                "created_at": str(r.get("_created_at", _now())),
                "updated_at": str(r.get("_created_at", _now())),
            }
            for r in rows
            if matches(r)
        ]
